### PR TITLE
feat(policies): add Wall-OSS-0.5 support

### PR DESCRIPTION
**Pull request by pkooij**
## Processor-owned tokenization update (2026-08-11)

- The serializable WALL-OSS-0.5 input pipeline now owns Qwen prompt construction, image processing, state discretization, tokenization, action-token masks, and sparse recipe text labels.
- `forward()`, `predict_action_chunk()`, and `generate_text()` consume those model-ready tensors and do not invoke an input tokenizer.
- Updated checkpoints: https://huggingface.co/pepijn223/walloss05-super-chatton-subtask/commit/fabe1f9e4d5f6d297b24033c8dba5f64521a94c8 and https://huggingface.co/lerobot/wall-oss-0.5/commit/eb780f31d4bd3e5abb87fdd102435a4d68e374d5
- Validation: 21 passed, 1 skipped; touched-file pre-commit hooks pass.

## Final #4183 contract (2026-08-11)

- WALL-OSS-0.5 explicitly returns True from supports_text_generation() and directly implements public generate_text(processed_batch).
- Its normal input pipeline starts with one RenderMessagesStep, then applies policy-owned WALL/Qwen multimodal formatting.
- Runtime is deliberately limited to #4380 vqa / next_subtask metadata; no runtime adapter or caller-controlled kind/prompt API remains.
- The built-in joint recipe lives in WallOSS05Config; recipe_path is only an explicit external override.
- Updated checkpoints: https://huggingface.co/pepijn223/walloss05-super-chatton-subtask/commit/4e51c27fd9a322afb90ff36f24082c542583f510 and https://huggingface.co/lerobot/wall-oss-0.5/commit/cc7553a33de60211c7b32bffe72e684a0978269a
- Focused final suite: 14 passed, 8 skipped; all touched-file pre-commit hooks passed.

## Summary

Wall-OSS-0.5 is a 4B vision-language-action model built from a Qwen2.5-VL vision-language backbone and a continuous flow-matching action expert. It is a different model generation from the existing `wall_x` / `wall-oss-flow` policy and is therefore registered separately as `wall_oss_05`.

This PR implements Wall-OSS-0.5 natively in LeRobot with PyTorch and Transformers. Transformers supplies the tokenizer, image processor, multimodal rotary embeddings, vision encoder, and tied language vocabulary. LeRobot implements the Wall-specific routed language/action experts, state serialization, text decoding, x-prediction flow objective, 10-step Euler sampler, and q01/q99 normalization through standard policy processors. No Wall-X checkout, runtime `PYTHONPATH`, PEFT, Accelerate, torchdiffeq, or qwen-vl-utils dependency is required.

The Qwen processor, checkpoint loading, prompt construction, multimodal model-input preparation, text/VQA generation, and action inference are owned directly by `WallOSS05Policy`; there is no separate Wall runtime wrapper.

WALL-OSS-0.5 exposes its native text head through public generate_text(processed_batch). The shared runtime routes contract-rendered VQA and next-subtask requests through the ordinary policy preprocessor, and generated subtasks condition following flow chunks.

This PR is stacked on and depends on #4183 (`codex/language-runtime`) for the shared interactive language runtime, recipe representation, and direct policy text API.

## Training

Training is supported through the standard `lerobot-train` path. Loading the base checkpoint for fine-tuning replaces its saved processor statistics with the training dataset statistics, and the fitted statistics are persisted with the new checkpoint.

The policy additionally supports recipe-driven language supervision:

- `recipe_path=None` preserves the original action-only flow path.
- A targeted assistant turn receives sparse next-token cross-entropy through Wall's tied Qwen language head.
- A `low_level` turn receives continuous flow-action supervision.
- A `low_level` recipe with a targeted assistant turn trains text and actions jointly in one causal sequence, so the action queries attend to the supervised text.

Only the targeted assistant payload and its `<|im_end|>` token are language labels; user prompts, padding, images, proprioception, and continuous action-query tokens remain masked. Text and flow losses default to the Wall report's relative weighting (`0.01` and `1.0`). Recipe rendering stays lazy so a saved language-trained processor can still load for ordinary inference with the Transformers-only Wall extra.

## Converted checkpoint

| Variant | LeRobot checkpoint | Intended use |
| --- | --- | --- |
| Base 4B | [`lerobot/wall-oss-0.5`](https://huggingface.co/lerobot/wall-oss-0.5) | Canonical 26D fine-tuning and deployment |

The published checkpoint includes the LeRobot config with its embedded Wall/Qwen `vlm_config`, Qwen processor/tokenizer assets, model weights, saved LeRobot pre/postprocessors and q01/q99 statistics, and conversion report. The legacy Wall training/config files and `.pth` normalizer files are not required or published.

## Model inputs and outputs

| Inputs | Internal representation | Output |
| --- | --- | --- |
| Task/question text; `face_view` + `right_wrist_view`; canonical 26D state | LeRobot q01/q99 processing and Wall's routed Qwen/action experts | 32 × 26 continuous actions, or autoregressive text for VQA, grounding, captions, and subtasks |

The current integration deliberately requires canonical 26D state and action features. Robot-specific action-space adapters are outside this PR.

## Author parity

Wall-X and LeRobot were run independently in BF16 on an RTX 5090. The comparison starts from raw state, action, camera, and task inputs and passes the LeRobot side through its saved preprocessor.

| Check | Result |
| --- | ---: |
| State and action normalization | Exact |
| Token IDs, attention mask, and processed pixels | Exact |
| `forward()` absolute loss difference | `0.0003154874` |
| First two greedy text token IDs | Exact: `[408, 198]` |

A compact action reproduction script and output are included in a PR comment. Text parity additionally compares the audited and native vocabulary logits on identical multimodal inputs; both select the same tested autoregressive tokens.

## Validation

- Focused WALL-OSS-0.5 contract suite — 14 passed, 8 skipped.
- Pre-commit hooks pass on all final touched files.

## Checklist

- [x] Native LeRobot policy without an internal or external Wall runtime wrapper
- [x] Public converted checkpoint with serialized config and processor-owned normalization statistics
- [x] Canonical 26D state/action path
- [x] Standard LeRobot action-only training with dataset-stat replacement
- [x] Recipe-driven text-only and joint text/action training
- [x] Native VQA and subtask generation through the direct policy API
- [x] Shared language-runtime contract and direct policy API integration
- [x] Strict checkpoint load, training forward/backward, and inference
- [x] End-to-end author/native parity rerun on RTX 5090
- [x] Focused policy, recipe-processor, and runtime tests
- [x] Targeted formatting, lint, typing, security, and documentation hooks pass
- [ ] Remaining CI jobs complete
- [ ] Community review linked

### CODE REVIEW

**Comment by pkooij** [comment]
Re-ran parity on the RTX 5090 after moving q01/q99 normalization into the LeRobot processors. This is the compact comparison core (the surrounding Wall-X SDPA import shim only makes the audited author checkout importable on this machine):

```python
AUTHOR = "/home/croissant/wall-oss-05-audit"
LEROBOT = "/home/croissant/wall-oss-05-processor-parity"
SEED = 1234

# `author` is the Wall-X adapter built from AUTHOR. Start both paths from raw values.
image = np.arange(448 * 448 * 3, dtype=np.uint8).reshape(448, 448, 3)
state = np.linspace(-0.5, 0.5, 26, dtype=np.float32)[None, None]
actions = torch.linspace(-0.75, 0.75, 32 * 26).reshape(1, 32, 26)
observation = {
    "proprioception": state,
    "agent_pos_mask": np.ones((1, 1, 26), dtype=bool),
    "dof_mask": np.ones((1, 32, 26), dtype=bool),
    "face_view": image,
    "right_wrist_view": np.flip(image, axis=1).copy(),
}
prefix, postfix = author._get_flow_prompt("pick up the object")
author_inputs = author._construct_model_input([observation], [prefix], [postfix])

config = WallOSS05Config.from_pretrained(LEROBOT)
pre, _ = make_pre_post_processors(config, pretrained_path=LEROBOT)
batch = pre({
    OBS_STATE: torch.from_numpy(state[:, 0]).clone(),
    ACTION: actions,
    "observation.images.face_view": torch.from_numpy(image.copy()).permute(2, 0, 1).float()[None] / 255,
    "observation.images.right_wrist_view": torch.from_numpy(observation["right_wrist_view"]).permute(2, 0, 1).float()[None] / 255,
    "task": ["pick up the object"],
})
torch.testing.assert_close(
    batch[OBS_STATE],
    author._normalizer_propri.normalize_data(torch.from_numpy(state), ["x2_normal"]).cuda()[:, 0],
)
author_actions = author._normalizer_action.normalize_data(actions, ["x2_normal"]).cuda()
torch.testing.assert_close(batch[ACTION], author_actions)

t = torch.tensor([0.2], device="cuda")
mask = torch.ones((1, 32), device="cuda", dtype=torch.bool)
torch.manual_seed(SEED); torch.cuda.manual_seed_all(SEED)
author_loss = author._model(
    **author_inputs, action_chunk=author_actions, sample_time=t, flow_loss_mask=mask
).loss.float()
shared = {k: author_inputs[k].cpu() for k in (
    "input_ids", "attention_mask", "pixel_values", "image_grid_thw", "moe_token_types"
)}
del author
gc.collect(); torch.cuda.empty_cache()

policy = WallOSS05Policy.from_pretrained(LEROBOT, config=config)
native_inputs, *_ = policy._build_model_inputs({
    k: v.cpu() if isinstance(v, torch.Tensor) else v for k, v in batch.items()
})
for key, expected in shared.items():
    torch.testing.assert_close(native_inputs[key].cpu(), expected)
torch.manual_seed(SEED); torch.cuda.manual_seed_all(SEED)
native_loss = policy.model(
    **native_inputs,
    action_chunk=batch[ACTION],
    sample_time=t,
    flow_loss_mask=mask,
).loss.float()
print(f"{abs(author_loss.item() - native_loss.item()):.10f}")
```

Result: `forward()` absolute loss difference = **`0.0003154874`**.

**Comment by suolyer** [comment]
Thanks for this thorough and clean port!

I have reviewed the code and verified the parity on our end. We also ran evaluations on our internal benchmark tasks using this LeRobot implementation, and the performance matches our original Wall-X framework.

Everything looks great to me. LGTM! Ready to merge. 🚀
